pkg/apis/build/v1alpha1/build_source.go

```diff
@@ -4,15 +4,66 @@
 
 package v1alpha1
 
+import corev1 "k8s.io/api/core/v1"
+
 // BuildSourceType enumerates build source type names.
 type BuildSourceType string
 
+const (
+	BuildSourceTypeGit  = "Git"
+	BuildSourceTypeHTTP = "HTTP"
+)
+
 // BuildSource remote artifact definition, also known as "sources". Simple "name" and "url" pairs,
 // initially without "credentials" (authentication) support yet.
 type BuildSource struct {
 	// Name instance entry.
 	Name string `json:"name"`
 
-	// URL remote artifact location.
+	// Type is the type of build source. Must be one of the following:
+	// - "Git"
+	// - "HTTP"
+	// TODO - populate this field on admission with a webhook
+	//
+	// +optional
+	Type BuildSourceType `json:"type,omitempty"`
```

Comment:
We should have some validation, I think. So that if the type is set to Git, that the Git section and only the Git section is provided.

Reply:
Yes - for this to be really effective we need a mutating admission webhook. We shouldn't expect users to set the Type, especially if we add OpenAPI `oneOf` validations to ensure only one of the sub-types are set.

Unfortunately kubebuilder doesn't support this via annotation today :disappointed:

Reply:
Maybe we should remove the `type` field completely then ? It would kind of by duplicate information:

```yaml
spec:
  sources:
    - name: my-source
      type: HTTP
      http:
        url: https://fromsomewheretoloadsomething/my-image.jpg
```

This would be shorter and still unambiguous:

```yaml
spec:
  sources:
    - name: my-source
      http:
        url: https://fromsomewheretoloadsomething/my-image.jpg
```